### DISCUSSION TITLE
Further explanation for Short Palindrome

### DISCUSSION THREAD

**Discussion by babhishek21** [babhishek21/oj-sols#3] [Q&A]
Refer to 2808e3c. (for [Short Palindrome (HackerRank World CodeSprint 5)](https://github.com/babhishek21/oj-sols/blob/2808e3c02123a2b040cd8ef2ede039104b27a5ad/misc/short_palindrome.cpp) )

Ask for any further explanations here.

**Comment by chinhodado** [comment]
Can you explain `dp_left` and `dp_right` in more details please?
Also looking at [this](https://github.com/babhishek21/oj-sols/blob/master/misc/short_palindrome.cpp#L95-L97) it seems you're counting `YXXY` and not `XYYX` like the comment on line 86 says?

**Comment by babhishek21** [comment]
### Further explanation for `dp_left` and `dp_right`

## 

The idea of the solution is to find _4-tuples_ of unique indexes **(a, b, c, d)** such that the characters at these positions are **X, Y, Y,** and **X** respectively. (NOTE: **X** and **Y** are variables and not actual literals.) Additionally **a < b < c < d** should be maintained.

Formally the set of such characters will form a valid _countable_ palindromic subsequence and hence can be represented as `X_a Y_b Y_c X_d`. 

Here the representation `X_a` denotes the set of characters which are constrained by the predicate `a`. For example, for the string `aabcbba`,
- `X_3` would mean character **'c'** which is at index 3
- `X_i>=4` would mean characters at indexes **>=4** which would mean the set **{ 'b', 'b', 'a' }**
- `X_3, X_4` would not mean _anything_ since the characters at indexes 3 and 4 are **_not**_ the same.

Finding such tuples naively can go up to quadratic time. We can however subdivide the problem to two things:
1. Find tuples 'X_a Y_b' (these are stored in `dp_right[][]`)
2. Find tuples 'Y_c X_d' (these are stored in `dp_left[][]`)

#### How exactly are these two arrays generated?

The idea is to store such occurrences based on the variable character **Y**. The reason is because, we can always choose any start and end positions for our valid palindromic subsequence by choosing _any two unique_ indexes for the variable character **X**.

`dp_left[i][Y]` refers to the count of characters formalised as `Y_c=i-1` for `X_d>=i` for the index `i`. An example will clear this up. Take the sample string `abbaab`. For index `i=3`,
- `Y_c=i-1` would refer to character **'b'** at index `i-1=2`.
- `X_d>=i` would refer to the set **{ 'a', 'a', 'b' }**.
  Since we can get 3 sets of `Y_c X_d` from this, namely **"ba"** (`c=2, d=3`), **"ba"** (`c=2, d=4`) and **"bb"** (`c=2, d=5`), the character **Y = 'b'** at index 2 will contribute 1 to `dp[3]['b']`, `dp[4]['b']` and `dp[5]['b']`.

For each of the `Y_c X_d`', if we could just find a set `X_a Y_b`, we would be all set with a valid palindromic subsequence. That is facilitated by `dp_right[i][Y]`.

How `dp_left[][]` and `dp_right[][]` are generated is a matter of finding the right recurrence. That can be easily inferred from the code.

**Comment by chinhodado** [comment]
Thank you for the explanation. So, in short, `dp_left[i][C]` is the count of character C to the left of index `i` in the string?

**Comment by babhishek21** [comment]
@chinhodado Yes indeed.

**Comment by pratiksatapathy** [comment]
`for(int i = nx-2; i>=0; i--) {
        lterm += dp_right[pos[x][i+1]][y]; 
        ans += lterm * dp_left[pos[x][i]][y]; 
        ans %= MOD;
      }`
Loved the way you have written this..
